main.go

```diff
@@ -56,76 +56,128 @@ var (
-		panic(fmt.Errorf("can't get instance"))
+	fmt.Printf("Using project: %s\n", project)
+
+	// Get instance ID
+	instance := instanceOverride
+	if instance == "" {
+		instance, err = getInstance(project)
+		if err != nil {
+			fmt.Printf("\nCan't get instance: %v\n", err)
+			panic(fmt.Errorf("can't get instance"))
+		}
 	}
+	fmt.Printf("Using Spanner instance: %s\n", instance)
 	printPermissionsWarning()
-	f, n, err := getSeekable(os.Stdin)
-	if err != nil {
-		printSeekError(err)
-		panic(fmt.Errorf("can't get seekable input file"))
-	}
-	defer f.Close()
-	conv := postgres.MakeConv()
-	dbName, err := getDatabaseName(now)
-	if err != nil {
-		fmt.Printf("\nCan't get database name: %v\n", err)
-		panic(fmt.Errorf("can't get database name"))
+
+	now := time.Now()
+	// Get database name
+	dbName := dbNameOverride
+	if dbName == "" {
+		dbName, err = getDatabaseName(now)
+		if err != nil {
+			fmt.Printf("\nCan't get database name: %v\n", err)
+			panic(fmt.Errorf("can't get database name"))
+		}
 	}
+
 	// If filePrefix not explicitly set, use dbName.
 	if filePrefix == "" {
 		filePrefix = dbName + "."
 	}
+
+	err = process(project, instance, dbName, nil, filePrefix, now)
+	if err != nil {
+		panic(err)
+	}
+}
+
+type ioStreams struct {
+	in, out *os.File
+}
+
+var ioHelper = &ioStreams{os.Stdin, os.Stdout}
+
+func process(projectID, instanceID, dbName string, helper *ioStreams, ouputFilePrefix string, now time.Time) error {
+	if helper != nil {
+		ioHelper = helper
+	}
+
+	f, n, err := getSeekable(ioHelper.in)
+	if err != nil {
+		printSeekError(err)
+		return fmt.Errorf("can't get seekable input file")
+	}
+	defer f.Close()
+	conv := postgres.MakeConv()
+
 	err = firstPass(f, n, conv)
 	if err != nil {
-		panic(fmt.Errorf("Failed to parse the data file: %v", err))
+		fmt.Fprintf(helper.out, "Failed to parse the data file: %v", err)
+		return fmt.Errorf("failed to parse the data file")
 	}
-	writeSchemaFile(conv, now, filePrefix+schemaFile)
-	db, err := createDatabase(project, instance, dbName, now, conv)
+	writeSchemaFile(conv, now, ouputFilePrefix+schemaFile)
+	db, err := createDatabase(projectID, instanceID, dbName, conv)
 	if err != nil {
 		fmt.Printf("\nCan't create database: %v\n", err)
-		panic(fmt.Errorf("can't create database"))
+		return fmt.Errorf("can't create database")
 	}
 	client, err := getClient(db)
 	if err != nil {
 		fmt.Printf("\nCan't create client for db %s: %v\n", db, err)
-		panic(fmt.Errorf("can't create Spanner client"))
+		return fmt.Errorf("can't create Spanner client")
 	}
 	_, err = f.Seek(0, 0)
 	if err != nil {
 		fmt.Printf("\nCan't seek to start of file (preparation for second pass): %v\n", err)
-		panic(fmt.Errorf("can't seek to start of file"))
+		return fmt.Errorf("can't seek to start of file")
 	}
 	rows := conv.Rows()
 	bw := secondPass(f, client, conv, rows)
-	report(bw, n, now, db, conv)
+	report(bw, n, now, db, conv, ouputFilePrefix+reportFile, ouputFilePrefix+badDataFile)
```

Comment:
Just checking: your motivation for passing these parameters is so that we can later refactor process into a separate module, and the parameters will capture everything we need from main.

Reply:
Yes. Later, when we move `process` into a seperate module, adding these parameters makes this method to be more testable (relying on function parameters instead of global variables).